### ISSUE TITLE
numpy.ndarray binary incompatibility error in environments with `numpy<1.20`

### ISSUE THREAD

**Issue by bosonie** [zerothi/sisl#310]
After upgrading sisl to 0.11.0 I can not import the package anymore.

The call:
``from sisl._indices import indices_le``
returns:
``ValueError: numpy.ndarray size changed, may indicate binary incompatibility. Expected 88 from C header, got 80 from PyObject``

OS Fedora 31
Python 3.7.6
numpy 1.17.5 (and others, see below)

Tried to reinstall numpy (same version) as suggested on the web, but couldn't solve the problem.
I am bound to this old numpy version due to other packages requirement, but I also tried in a new environment (same python version) numpy 1.18 and 1.19. Got the same error. Version of numpy 1.20, on the contrary, works.

**Comment by zerothi** [comment]
Thanks, so are you using conda? Because installing sisl should use your local numpy version (there are no hard-coded numpy versions).

If you build `sisl` locally it uses your currently installed numpy version?

**Comment by bosonie** [comment]
I'm using `virtualenv`. The new environment has no `numpy` and therefore it is installed as a dependence of sisl (numpy>=1.13 is the requirement).
In any other environment (for example the local env where I have numpy 1.17.5), it uses the installed numpy.
I tried both installing with PyPI or locally after `git clone`. The result is the same. Unless U=I have numpy 1.20, I get the same error.

**Comment by zerothi** [comment]
It really sounds like the build is using another numpy version from your installed one.

There is _nothing_ hard-coded regarding the things you mention, the pyx files are cythonized at build time against your numpy + cython and compiler.

Please give more details
- build output
- module versions + paths as printed once loaded in python

**Comment by bosonie** [comment]
I'm not sure what you require. But let's start from the very basic. I have a new `virtualenv` called `www`. I first install numpy 1.17.5 (pip install numpy==1.17.5) and then I do `pip install sisl`. The output is:
```
Collecting sisl
  Downloading sisl-0.11.0.tar.gz (4.1 MB)
     |████████████████████████████████| 4.1 MB 1.7 MB/s 
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
    Preparing wheel metadata ... done
Requirement already satisfied: setuptools in ./lib/python3.7/site-packages (from sisl) (54.0.0)
Collecting netCDF4
  Downloading netCDF4-1.5.6-cp37-cp37m-manylinux2014_x86_64.whl (4.7 MB)
     |████████████████████████████████| 4.7 MB 15.5 MB/s 
Collecting pyparsing>=1.5.7
  Downloading pyparsing-2.4.7-py2.py3-none-any.whl (67 kB)
     |████████████████████████████████| 67 kB 40.8 MB/s 
Requirement already satisfied: numpy>=1.13 in ./lib/python3.7/site-packages (from sisl) (1.17.5)
Collecting scipy
  Downloading scipy-1.6.1-cp37-cp37m-manylinux1_x86_64.whl (27.4 MB)
     |████████████████████████████████| 27.4 MB 33.6 MB/s 
Collecting cftime
  Downloading cftime-1.4.1-cp37-cp37m-manylinux2014_x86_64.whl (313 kB)
     |████████████████████████████████| 313 kB 45.7 MB/s 
Building wheels for collected packages: sisl
  Building wheel for sisl (PEP 517) ... done
  Created wheel for sisl: filename=sisl-0.11.0-cp37-cp37m-linux_x86_64.whl size=14890322 sha256=24034d1cb6829a62bb102068ae144a5e32e73c18eed8a9aff9e5f3dc514fcd5b
  Stored in directory: /tmp/pip-ephem-wheel-cache-nn7elgn0/wheels/b6/b8/e1/a5f12406cf689dd1379383ab242c4424f9477a03406eebbfea
Successfully built sisl
Installing collected packages: cftime, scipy, pyparsing, netCDF4, sisl
Successfully installed cftime-1.4.1 netCDF4-1.5.6 pyparsing-2.4.7 scipy-1.6.1 sisl-0.11.0
```
I can not load sisl since the call that crashes is in the `__init__()`. However the location of the package seems ok!
```
>>> numpy.__file__
'/home/ebosoni/www/lib/python3.7/site-packages/numpy/__init__.py'
```
Error:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/ebosoni/www/lib/python3.7/site-packages/sisl/__init__.py", line 90, in <module>
    from . import utils
  File "/home/ebosoni/www/lib/python3.7/site-packages/sisl/utils/__init__.py", line 31, in <module>
    from . import mathematics as math
  File "/home/ebosoni/www/lib/python3.7/site-packages/sisl/utils/mathematics.py", line 7, in <module>
    from sisl._indices import indices_le
  File "sisl/_indices.pyx", line 1, in init sisl._indices
ValueError: numpy.ndarray size changed, may indicate binary incompatibility. Expected 88 from C header, got 80 from PyObject
```
Do you need other distribution info?

**Comment by bosonie** [comment]
As an addition, I also tried to install with `--no-cache-dir` option for both sisl and numpy but the outcome is the same.

**Comment by zerothi** [comment]
Could you build it yourself and give me the output?

**Comment by zerothi** [comment]
As per https://stackoverflow.com/q/66060487/827281 this really boils down to incompatible versions. So the ones that are compiled against each other are not consistent. Did you try just importing scipy?

**Comment by bosonie** [comment]
I built from source (git clone) and `pip install -e .`:
```
Obtaining file:///home/ebosoni/qqq/sisl
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
    Preparing wheel metadata ... done
Collecting scipy
  Downloading scipy-1.6.1-cp37-cp37m-manylinux1_x86_64.whl (27.4 MB)
     |████████████████████████████████| 27.4 MB 15.9 MB/s 
Collecting pyparsing>=1.5.7
  Downloading pyparsing-2.4.7-py2.py3-none-any.whl (67 kB)
     |████████████████████████████████| 67 kB 14.1 MB/s 
Requirement already satisfied: setuptools in /home/ebosoni/qqq/lib/python3.7/site-packages (from sisl==0.11.0.dev0) (54.0.0)
Collecting netCDF4
  Downloading netCDF4-1.5.6-cp37-cp37m-manylinux2014_x86_64.whl (4.7 MB)
     |████████████████████████████████| 4.7 MB 10.9 MB/s 
Requirement already satisfied: numpy>=1.13 in /home/ebosoni/qqq/lib/python3.7/site-packages (from sisl==0.11.0.dev0) (1.17.5)
Collecting cftime
  Downloading cftime-1.4.1-cp37-cp37m-manylinux2014_x86_64.whl (313 kB)
     |████████████████████████████████| 313 kB 2.7 MB/s 
Installing collected packages: cftime, scipy, pyparsing, netCDF4, sisl
  Running setup.py develop for sisl
Successfully installed cftime-1.4.1 netCDF4-1.5.6 pyparsing-2.4.7 scipy-1.6.1 sisl
```
Error persist

Scipy imports without problems. Also I want to remind that sisl version 0.10.0 works perfectly fine.

**Comment by zerothi** [comment]
No, I want the complete build information. `python setup.py install.... `

Yes, but all I can find shows that this is due to non same installations of numpy and the used package. In this case sisl. So it boils down to figuring out how numpy was build, and do the same with sisl. There are no hard coded sizes, all is determined at compile time.

**Comment by zerothi** [comment]
A guess could be that your cython installation has a hard coded numpy link (as may happen with easy install paths.)

**Comment by zerothi** [comment]
This would show it self by comparing building 0.10.0 and 0.11.0

**Comment by bosonie** [comment]
Probably going in the right direction.

Since now, all the cases I reported where in an environment where Cython was not installed.

Now I pip installed it (after numpy==1.17.5) and built (`python setup.py install`).
The error is now disappeared. I attach the out and log.

[out.txt](https://github.com/zerothi/sisl/files/6085402/out.txt)
[log.txt](https://github.com/zerothi/sisl/files/6085403/log.txt)

In the setup.py I see quite a bit of logic if Cython is not present. Is it necessary for the the installation? If so, was Cython took from somewhere else in previous cases? That would make sense

**Comment by pfebrer** [comment]
Hi, this happened to a collaborator of mine as well. Since she had no attachment to the environment (in this case @bosonie  needs to stick with it because it has aiida installed in it) I advised her to just create a new environment, which worked, I guess that installed the most recent version of numpy.

I can reproduce the issue by following this steps:
- Create new environment
- `pip install numpy==1.17.5 Cython`
- `git clone https://github.com/zerothi/sisl.git`
- `cd sisl`
- `pip install .`

By looking at the log, I see the following line:

```
  Successfully installed Cython-0.29.22 numpy-1.20.1 setuptools-54.0.0 wheel-0.36.2
```

, so somehow it installs numpy 1.20.1 even though I have 1.17.5. The weirdest part is that after the build, if I check the version of numpy in IPython, it says it is 1.17.5. So it uses 1.20.1 just to build sisl, I don't know why.

Installing it with `python setup.py install` works because it doesn't manage the dependencies and therefore it doesn't use a different numpy.

**Comment by pfebrer** [comment]
Basically my comment is the same as @bosonie's hahaha

**Comment by pfebrer** [comment]
> In the setup.py I see quite a bit of logic if Cython is not present. Is it necessary for the installation? If so, was Cython took from somewhere else in previous cases? That would make sense

You are probably able to install it without` cython` because you are not using a new sisl folder, therefore all the files have been generated in the previous attempts when you did have `cython` and you don't need `cython` for further attempts.

**Comment by bosonie** [comment]
Thanks @pfebrer for jumping in.
I believe I used a new sisl folder in a new virtual environment. `pip install .` was not complaining about missing `cython`, only `python setup.py install` does.

**Comment by bosonie** [comment]
In other words:
* Create new environment
* `pip install numpy==1.17.5`
* `git clone https://github.com/zerothi/sisl.git`
* `cd sisl`
* `pip install .`

Meaning exactly like Pol's steps but without cython, it is even worse, because it does not even show `Successfully installed  numpy-1.20.1` but apparently uses something weird.

**Comment by bosonie** [comment]
Update.
I found out my sisl 0.10.0 installation was working only because it was getting some cached wheels somewhere. Therefore the problem is present for version 0.10.0 as well (I'll change the title).
I believe that at the time I installed sisl for the first time, numpy 1.20 was not out yet and therefore the fact that sisl was compiled with the newest version of numpy was not a problem.
However, probably, since numpy 1.20 we have some changes (maybe here https://numpy.org/doc/stable/release/1.20.0-notes.html#c-api-changes) and the fact that sisl uses the newest version of numpy is causing incompatibility problems.
It would be great to have feedback from other environments and python versions to understand if the problem that @pfebrer and I see is widespread. Moreover, it is left to understand why the pip installation uses a new version of numpy to install sisl.
In any case the solution is to `python setup.py install`

**Comment by zerothi** [comment]
Ok, I think the problem comes down to Cython not being present at installation time.

Could you try with pip install with the latest commit? That should force-install Cython, I didn't have it as a dependency which seems to be the problem. Repeat with a fresh environment without Cython.

1. Cython is necessary to build the `.c` sources from the `.pyx` files. I have previously had them shipped in this repo. But I think that making the installation process create them would be more stable since then the user isn't depending on my installed versions
2. sisl doesn't upload wheels. So I guess all versions that didn't have the C sources shipped would suffer from this problem.
3. Indeed this seems related to the 1.20 update. Remember, sisl doesn't have a fixed numpy requirement (just above 1.13). 
4. As far as I understand this comes down to not having Cython installed. But lets see.
5. `pip` installation process is *defaulting* to build in a sand-box mode. To bypass this one can do `pip install --no-deps`.

**Comment by pfebrer** [comment]
> pip installation process is defaulting to build in a sand-box mode. To bypass this one can do pip install --no-deps.

What happens if you build with `numpy 1.13` and then you use `sisl` with `numpy 1.20`? Does that work? If so, shouldn't it be specified that `sisl` has to be built with the oldest version of `numpy` supported?

Otherwise anyone that tries to install sisl with `pip` in an old environment will encounter this problem. (not a good thing, is it? :( )

**Comment by zerothi** [comment]
> > pip installation process is defaulting to build in a sand-box mode. To bypass this one can do pip install --no-deps.
> 
> What happens if you build with `numpy 1.13` and then you use `sisl` with `numpy 1.20`? Does that work? If so, shouldn't it be specified that `sisl` has to be built with the oldest version of `numpy` supported?
> 
> Otherwise anyone that tries to install sisl with `pip` in an old environment will encounter this problem. (not a good thing, is it? :( )

Well, probably upgrading `numpy` should solve this.

But there is no *way* around adjusting older sisl installations. The problem is that I don't upload wheels, and thus things need to be built upon installation.
If any user requests an older sisl version they may encounter this. My bet is that there are quite a bit of packages that will suffer from this change.

Secondly, I don't want to restrict users from using numpy 1.20. sisl does support this, and if I say that sisl doesn't support this, then we'll have more problems when users start using that version. Because then one cannot install sisl.

This is the first time I have heard of this problem, and it seems like it is quite easy to fix, 1) ensure cython is installed and 2) ensure cython+sisl installation uses your production numpy version.

I am open to suggestions, but currently I don't see an easy solution that is forward and backward compatible?

**Comment by bosonie** [comment]
I'm sad to report that the introduction of `Cython` as a direct dependence, does not solve the issue.
I took particular care to make sure that no cached package is used, but if in this case the problem persist. I do not know what to say.
I just add that `virtualenv` install `setuptools`, `pip` and  `wheel` when a virual environment is created.
I'm now exploring what happens with the `--no-deps` option

**Comment by zerothi** [comment]
Damn... I'll have to think some more...

**Comment by pfebrer** [comment]
Nick, this problem should be solved by publishing the wheels (built with `numpy 1.13`), right? Maybe that is worth it to avoid people having installation problems (and finally making sisl quick to install through pip :smile: )

**Comment by bosonie** [comment]
I link here the most complete tread I found on the problem.
scikit-learn-contrib/hdbscan#457
I'll read through it later

**Comment by zerothi** [comment]
Yeah, I am now pretty sure this is because of mixing compilation numpy versions with running numpy versions. And it seems this issue is starting to appear all over. :)

**Comment by zerothi** [comment]
I have worked a bit on this. And to me this can be fixed by doing:
```shell
# ensure everything is pre-installed!
pip install -r requirements.txt 
pip install --no-build-isolation sisl
```
this forces the building numpy to be the one you have installed.
Could you report if this works?

**Comment by zerothi** [comment]
> > Nick, this problem should be solved by publishing the wheels (built with `numpy 1.13`), right? Maybe that is worth it to avoid people having installation problems (and finally making `sisl` quick to install through `pip`  )
> 
> Yes, indeed this should be the case. I am running some test examples to make an easy work-around for users.
> 
> However, it would require wheels for both _pre_ 1.20 versions and _post_ 1.20 versions so it can work in both cases.

Actually, I am not sure this is workable all around. It requires that the wheel is built against a numpy version that is backwards compatible, and 1.20 seems not to be this.

Details about the problem is actually seen here:
- https://numpy.org/doc/stable/release/1.20.0-notes.html#size-of-np-ndarray-and-np-void-changed
- https://numpy.org/doc/stable/release/1.20.0-notes.html#the-pyarray-descrcheck-macro-is-modified

A proper documentation on what to do is here: https://numpy.org/devdocs/user/depending_on_numpy.html#build-time-dependency

Basically this is occurring all over the eco-system.
What `scipy` suggests is to do what I did in 7cc0e584ad68b442bdaaf1c7b69511cfd8ad957a which I think/hope/wish fixes this alltogether. This should also mean that one does not need to use the `--no-build-isolation` flag.

**Comment by pfebrer** [comment]
If I understand correctly, this is what I was proposing moreless. Numpy versions are not backwards compatible, but they are all forward compatible.

That's why I was proposing to build the wheels with the oldest numpy supported (`1.13`). Then it should work for whatever version of numpy the user has in their environment. To my understanding, the solution you propose is doing this but with two differences:
- You also made sure that when users compile from source they use an old numpy version (not only when we build the wheels).
- The `oldest-supported-numpy` metapackage also considers if the version is compatible with the python version.

Sounds good to me!

The only doubt I have is: does `oldest-supported-numpy` consider that sisl only supports `>=1.13` or can it try to use an even older version? Would it matter if it picked an older version? I.e., is the requirement for `>=1.13` a build time requirement or just a requirement for usage (say because a specific method was introduced in `1.13`)?

**Comment by zerothi** [comment]
Yes, you are correct, it is basically the same as you suggest.

> The only doubt I have is: does `oldest-supported-numpy` consider that sisl only supports `>=1.13` or can it try to use an even older version? Would it matter if it picked an older version? I.e., is the requirement for `>=1.13` a build time requirement or just a requirement for usage (say because a specific method was introduced in `1.13`)?

I think I just fixed this. sisl only has a build requirement for the oldest numpy version available for the current python version. And once runned it requires a specific version. I just committed f364af06be2aa1854d8205bcad24c84efc598db8 which removed my extraneous requirement in `requires`! That variable is only a _build_ requirement.

**Comment by pfebrer** [comment]
I'll try to

```
pip install numpy==1.13
pip install /sisl
```

this afternoon. If it works, then problem solved, right?

**Comment by zerothi** [comment]
I have now fixed it and tested some cases. 

I am pretty sure (fingers crossed) that this is resolved for any later releases.

@pfebrer please have a look at the latest commit message and the file `developments/test_numpy_installs.sh` which tests a few things. You'll note that this one will fail for v0.11.0 (due to not having proper build-dependencies) but will work regardless in the latest commits.

There were some things that needed updating, but I think this is on the right track.

Again, for future reference, one should be able to install sisl without problems if you pre-install the dependencies, and then subsequently install sisl with the `--no-build-isolation` to forcefully use the currently used versions:
```shell
python3 -m pip install -r requirements.txt
python3 -m pip install --no-build-isolation sisl-0.11.0.tar.gz
```

I'll close this, please re-open if not fixed.
